modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/disk/SortedExternalResult.java

```diff
@@ -0,0 +1,473 @@
+/*
+ * Copyright 2019 GridGain Systems, Inc. and Contributors.
+ *
+ * Licensed under the GridGain Community Edition License (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.gridgain.com/products/software/community-edition/gridgain-community-edition-license
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.processors.query.h2.disk;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.LinkedList;
+import java.util.NoSuchElementException;
+import java.util.PriorityQueue;
+import java.util.Queue;
+import java.util.TreeMap;
+import org.apache.ignite.internal.GridKernalContext;
+import org.apache.ignite.internal.processors.query.h2.H2MemoryTracker;
+import org.apache.ignite.internal.processors.query.h2.H2Utils;
+import org.apache.ignite.internal.util.typedef.F;
+import org.apache.ignite.internal.util.typedef.internal.U;
+import org.h2.engine.Session;
+import org.h2.result.ResultExternal;
+import org.h2.result.SortOrder;
+import org.h2.store.Data;
+import org.h2.value.Value;
+import org.h2.value.ValueRow;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * This class is intended for spilling to the disk (disk offloading) sorted intermediate query results.
+ */
+@SuppressWarnings("MissortedModifiers")
+public class SortedExternalResult extends AbstractExternalResult {
+    /** Distinct flag. */
+    private final boolean distinct;
+
+    /** {@code DISTINCT ON(...)} expressions. */
+    private final int[] distinctIndexes;
+
+    /** Visible columns count. */
+    private final int visibleColCnt;
+
+    /** Sort order. */
+    private final SortOrder sort;
+
+    /** Last written to file position. */
+    private long lastWrittenPos;
+
+    /** In-memory buffer for gathering rows before spilling to disk. */
+    private TreeMap<ValueRow, Value[]> sortedRowsBuf;
+
+    /** In-memory buffer for gathering rows before spilling to disk. */
+    private ArrayList<Value[]> unsortedRowsBuf;
+
+    /** Sorted chunks addresses on the disk. */
+    private final Collection<Chunk> chunks;
+
+    /**
+     * Hash index for fast lookup of the distinct rows.
+     * RowKey -> Row address in the row store.
+     */
+    private ExternalResultHashIndex hashIdx;
+
+    /**
+     * Result queue.
+     */
+    private Queue<Chunk> resQueue;
+
+    /**
+     * Comparator for {@code sortedRowsBuf}.
+     */
+    private Comparator<Value> cmp;
```

Comment:
It would be great to leave a comment describing a difference between `cmp` and `sort` fields use. Both are Comparators  but as a reader I cannot figure out why do we need two.